compiler/hash-attrs/src/target.rs

```diff
@@ -0,0 +1,292 @@
+                AttrTarget::ModDef => allowed_argument_kinds.push("mod block"),
+                AttrTarget::Block => allowed_argument_kinds.push("body block"),
+                AttrTarget::Import => allowed_argument_kinds.push("import"),
+                AttrTarget::StructDef => allowed_argument_kinds.push("`struct` definition"),
+                AttrTarget::EnumDef => allowed_argument_kinds.push("`enum` definition"),
+                AttrTarget::TyFnDef => allowed_argument_kinds.push("type function definition"),
+                AttrTarget::FnDef => allowed_argument_kinds.push("`function` definition"),
+                AttrTarget::Ty => allowed_argument_kinds.push("type"),
+                AttrTarget::Expr => allowed_argument_kinds.push("expression"),
+                AttrTarget::Pat | AttrTarget::PatArg => allowed_argument_kinds.push("pattern"),
+                AttrTarget::TyArg => allowed_argument_kinds.push("type argument"),
+                AttrTarget::Field => allowed_argument_kinds.push("field"),
+                AttrTarget::EnumVariant => allowed_argument_kinds.push("enum variant"),
+                AttrTarget::MatchCase => allowed_argument_kinds.push("match case"),
+                _ => {}
+            }
+        }
+
+        write!(
+            f,
+            "{}",
+            SequenceDisplay::new(
+                &allowed_argument_kinds,
+                SequenceDisplayOptions {
+                    quote: false,
+                    mode: SequenceJoinMode::Either,
+                    ..Default::default()
+                }
+            )
+        )
+    }
+}
+
+/// The various kinds of AST nodes that an attribute can be applied to. The
+/// [AttrNode] is used to perform further validation checks on attribute
+/// applications after the initial parameter/subject checks have finished.
+#[derive(Clone, Copy, Debug)]
+pub enum AttrNode<'ast> {
+    /// A more general reference to an expression.
+    Expr(ast::AstNodeRef<'ast, ast::Expr>),
+
+    /// A function.
+    Fn(ast::AstNodeRef<'ast, ast::FnDef>),
+
+    /// A struct definition.
+    Struct(ast::AstNodeRef<'ast, ast::StructDef>),
+
+    /// An enum definition.
+    Enum(ast::AstNodeRef<'ast, ast::EnumDef>),
+
+    /// A enum definition variant.
+    EnumVariant(ast::AstNodeRef<'ast, ast::EnumDefEntry>),
+
+    /// A module definition.
+    Mod(ast::AstNodeRef<'ast, ast::ModDef>),
+
+    /// A match case.
+    MatchCase(ast::AstNodeRef<'ast, ast::MatchCase>),
+
+    /// A field or function parameter.
+    Param(ast::AstNodeRef<'ast, ast::Param>),
+
+    /// An argument to a constructor.
+    ExprArg(ast::AstNodeRef<'ast, ast::ExprArg>),
+
+    /// A pattern.
+    Pat(ast::AstNodeRef<'ast, ast::Pat>),
+
+    /// A pattern argument.
+    PatArg(ast::AstNodeRef<'ast, ast::PatArg>),
+
+    /// A type.
+    Ty(ast::AstNodeRef<'ast, ast::Ty>),
+
+    /// A type argument.
+    TyArg(ast::AstNodeRef<'ast, ast::TyArg>),
+}
+
+impl<'ast> AttrNode<'ast> {
```

Comment:
I guess this is the repetition you were trying to avoid...

Reply:
I think its kind of unavoidable here...